### ISSUE TITLE
`implied_bounds_in_impls` false positive

### ISSUE THREAD

**Issue by feds01** [rust-lang/rust-clippy#11435]
### Summary

The lint suggests removing a specified bound, however, removing the bound breaks code that then uses the function.

### Lint Name

implied_bounds_in_impls

### Reproducer

I tried this code:

```rust
use std::{
    cell::RefCell,
    fmt,
    ops::{Deref, DerefMut},
};

pub struct HeavyState<T: fmt::Debug> {
    current: RefCell<T>,
}

impl<T: fmt::Debug> HeavyState<T> {
    pub fn get_mut(&self) -> impl Deref<Target = T> + DerefMut + '_ {
        self.current.borrow_mut()
    }
}

pub fn foo(bindings: HeavyState<Vec<u32>>) {
   let mut b = bindings.get_mut();
   b.push(0);
}

fn main() {}
```

Clippy reports the following:

```
warning: this bound is already specified as the supertrait of `DerefMut`
  --> src/main.rs:14:35
   |
14 |     pub fn get_mut(&self) -> impl Deref<Target = T> + DerefMut + '_ {
   |                                   ^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#implied_bounds_in_impls
   = note: `#[warn(clippy::implied_bounds_in_impls)]` on by default
help: try removing this bound
   |
14 -     pub fn get_mut(&self) -> impl Deref<Target = T> + DerefMut + '_ {
14 +     pub fn get_mut(&self) -> impl DerefMut + '_ {
   |
```

However, applying the suggestion results in the following:
```
error[E0599]: no method named `push` found for opaque type `impl std::ops::DerefMut + '_` in the current scope
  --> src/main.rs:23:7
   |
23 |     b.push(0);
   |       ^^^^ method not found in `impl DerefMut`
```

The suggestion should be to transfer the associated type bound `Target = T` onto `DerefMut` rather than entirely removing the bound. This would be the correct code:
```rs
impl<T: fmt::Debug> HeavyState<T> {
    pub fn get_mut(&self) -> impl DerefMut<Target = T> + '_ {
        self.current.borrow_mut()
    }
}
```

### Version

```text
rustc 1.74.0-nightly (84a9f4c6e 2023-08-29)
binary: rustc
commit-hash: 84a9f4c6e6f0f02bff9acc9a5e0305a1554a87fc
commit-date: 2023-08-29
host: x86_64-apple-darwin
release: 1.74.0-nightly
LLVM version: 17.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by MingweiSamuel** [comment]
Bit simpler of an example, using iterators:
```rust
pub fn my_iter() -> impl Iterator<Item = u32> + DoubleEndedIterator {
    0..5
}
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e8cae5003e3b0bacea53972e1162d9c1

**Comment by y21** [comment]
Yeah, this lint needs some more love 😢 I created a PR that fixes a crash, as well as moving it to the nursery category for now, because there's a bunch of cases in which the suggestion it creates is invalid. Here's one more `I-suggestion-causes-error`:
```rs
use std::ops::{Deref,DerefMut};

fn x() -> impl DerefMut + Deref<Target = u8> { Box::new(1) }
```
```
help: try removing this bound
  |
3 - fn x() -> impl DerefMut + Deref<Target = u8> { Box::new(1) }
3 + fn x() -> impl DerefMut +  { Box::new(1) }
```
(note the trailing `+`)

This happens when the implied bound that is being removed is the last bound. Clippy already tries to extend the span if there's a `+` *ahead* of it, but not behind it

**Comment by MingweiSamuel** [comment]
Oh yeah, don't even need the type args for that syntactic error
```rust
pub fn x() -> impl Copy + Clone {
    5
}
```
```rust
help: try removing this bound
  |
1 - pub fn x() -> impl Copy + Clone {
1 + pub fn x() -> impl Copy +  {
  |
```
`MachineApplicable`was a bit ambitious :P
